File: tests/helics/application_api/CInterface/pi_sender.c
static char help[] = " PI SENDER: Simple program to demonstrate the usage of HELICS C Interface.\n\
            This example creates a ZMQ broker and a value federate.\n\
            The value federate creates a global pulications and publishes\n\
            t*pi for 20 time-steps with a time-step of 0.01 seconds.\n\n";


#include <stdio.h>
#include <ValueFederate_c.h>
#ifdef _MSC_VER
#include <windows.h>
#else
#include <unistd.h>
#endif

int main(int argc,char **argv)
{
  helics_federate_info_t fedinfo;
  const char*    helicsversion;
  helicsStatus   status;
  helics_broker  broker;
  const char*    initstring="2 --name=mainbroker";
  const char*    fedinitstring="--broker=mainbroker --federates=1";
  int            isconnected;
  double         deltat=0.01;
  helics_value_federate vfed;
  helics_publication pub;

  helicsversion = helicsGetVersion();

  printf("PI SENDER: Helics version = %s\n",helicsversion);
  printf("%s",help);

  /* Create broker */
  broker = helicsCreateBroker("zmq","",initstring);

  isconnected = helicsBrokerisConnected(broker);

  if(isconnected) {
    printf("PI SENDER: Broker created and connected\n");
  }
  
  /* Create Federate Info object that describes the federate properties */
  fedinfo = createFederateInfoObject();
  
  /* Set Federate name */
  status = FederateInfoSetFederateName(fedinfo,"TestA Federate");

  /* Set core type from string */
  status = FederateInfoSetCoreTypeFromString(fedinfo,"zmq");

  /* Federate init string */
  status = FederateInfoSetCoreInitString(fedinfo,fedinitstring);

  /* Set the message interval (timedelta) for federate. Note that
     HELICS minimum message time interval is 1 ns and by default
     it uses a time delta of 1 second. What is provided to the
     setTimedelta routine is a multiplier for the default timedelta.
  */
  /* Set one second message interval */
  status = FederateInfoSetTimeDelta(fedinfo,helicsTimeFromDouble(deltat));

  status = FederateInfoSetLoggingLevel(fedinfo,1);

  /* Create value federate */
  vfed = helicsCreateValueFederate(fedinfo);
  printf("PI SENDER: Value federate created\n");

  /* Register the publication */
  pub = helicsRegisterGlobalPublication(vfed,"testA","double","");
  printf("PI SENDER: Publication registered\n");

  /* Enter execution mode */
  status = helicsEnterExecutionMode(vfed);
  printf("PI SENDER: Entering execution mode\n");

  /* This federate will be publishing deltat*pi for numsteps steps */
  double this_time = 0.0;
  double value = 22.0/7.0,val;
  helics_time_t currenttime;
  int           numsteps=20,i;

  for(i=0; i < numsteps; i++) {
    val = i*deltat*value;

    currenttime = helicsRequestTime(vfed,helicsTimeFromDouble(this_time+deltat*i));

    printf("PI SENDER: Sending value %3.2fpi = %4.3f at time %3.2f to PI RECEIVER\n",deltat*i,val,this_time+deltat*i);
    status = helicsPublishDouble(pub,val);
  }

  status = helicsFinalize(vfed);
  printf("PI SENDER: Federate finalized\n");


  while(helicsBrokerisConnected(broker)) {
#ifdef _MSC_VER
	  Sleep(1);
#else
    usleep(1000); /* Sleep for 1 millisecond */
#endif
  }
  printf("PI SENDER: Broker disconnected\n");

  return(0);
}

File: docs/user-guide/examples/user_guide_examples/Example_3/pisender.py
# -*- coding: utf-8 -*-
import time
import sys

from math import pi
import helics as h

federate_name = f"SenderFederate{sys.argv[1]}"

print(f"{federate_name}: Helics version = {h.helicsGetVersion()}")

fedinfo = h.helicsCreateFederateInfo()

h.helicsFederateInfoSetCoreName(fedinfo, federate_name)
h.helicsFederateInfoSetCoreTypeFromString(fedinfo, "zmq")
h.helicsFederateInfoSetCoreInitString(fedinfo, "--federates=1")
h.helicsFederateInfoSetTimeProperty(fedinfo, h.helics_property_time_delta, 0.01)

vfed = h.helicsCreateValueFederate(federate_name, fedinfo)
print(f"{federate_name}: Value federate created")

pub = h.helicsFederateRegisterGlobalTypePublication(
    vfed, f"globaltopic{sys.argv[1]}", "double", ""
)
pub = h.helicsFederateRegisterTypePublication(
    vfed, f"localtopic{sys.argv[1]}", "double", ""
)

print(f"{federate_name}: Publication registered")

h.helicsFederateEnterExecutingMode(vfed)
print(f"{federate_name}: Entering execution mode")

this_time = 0.0
value = pi

for t in range(5, 10):
    val = value

    currenttime = h.helicsFederateRequestTime(vfed, t)

    h.helicsPublicationPublishDouble(pub, val)
    print(f"{federate_name}: Sending value pi = {val} at time {currenttime}")

    time.sleep(1)

h.helicsFederateFinalize(vfed)
print(f"{federate_name}: Federate finalized")

h.helicsFederateFree(vfed)
h.helicsCloseLibrary()
